docs/specs/ML.Net Sweepable API.md

```diff
@@ -0,0 +1,162 @@
+
+    // one of [true, false]
+    [Choice]
+    public bool UseSoftmax {get; set;}
+
+    // nested search space
+    [Option]
+    public Option AnotherOption {get;set;}
+}
+
+var ss = new SearchSpace<Option>();
+
+// each search space has a 1-d feature space where each feature is [0, 1). And search space will handle the mapping between hpo space and feature space so that tuner only needs to perform search on feature space, which both dimension and range are known.
+var parameter = ss.SampleFromFeatureSpace(new []{0,0,0,0,0,0});
+
+// auto-binding
+parameter.WindowSize.Should().Be(2);
+parameter.SeriesLength.Should().Be(2);
+parameter.UseSoftmax.Should().BeTrue();
+parameter.AnotherOption.WindowSize.Should().Be(2);
+
+// search space can also map parameter back to feature space
+ss.MappingToFeatureSpace(parameter).Should().BeEquivalantTo(0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
+```
+
+or correspondingly, via scratch
+``` csharp
+var ss = new SearchSpace();
+ss.Add("WindowSize", new UniformIntOption(2, 32768, true, 2));
+ss.Add("SeriesLength", new ChoiceOption(2,3,4));
+ss.Add("UseSoftmax", new ChoiceOption(true, false));
+ss.Add("AnotherOption", ss.Clone());
+
+var parameter = ss.SampleFromFeatureSpace(new []{0,0,0,0,0,0});
+
+// auto-binding doesn't exist for scratch api
+parameter["WindowSize"].AsType<int>.Should().Be(2);
+parameter["SeriesLength"].AsType<int>.Should().Be(2);
+parameter["UseSoftmax"].AsType<bool>.Should().BeTrue();
+parameter["AnotherOption"]["WindowSize"].AsType<int>.Should().Be(2);
+
+// search space can also map parameter back to feature space
+ss.MappingToFeatureSpace(parameter).Should().BeEquivalantTo(0.0, 0.0, 0.0, 0.0, 0.0, 0.0);
+```
+
+Currently, in order to make auto-binding work, there's a limitation on the parameter type that can be added to search space, which has to be either a Json primitive type or a nested search space.
+
+## sweepable estimator
+sweepable estimator allows user to combine search space with estimators in a similar way how ml.net estimator/pipeline are created. You use `CreateSweepableEstimator`, which accepts a lambda function and a search space to create a sweepable estimator. And it also provides `.Append` extension method so you can append sweepable estimator similiar with how you append other ml.net extimator. The bellow example presents how to create a pipeline with two sweepable estimators, one for text featurizor and the other for fast tree, for `titanic` dataset via `Sweepable()` extension.
+
+``` csharp
+var context = new MLContext();
+var fastTreeSS = new SearchSpace<FastTreeOption>();
+var textFeaturizeSS = new SearchSpace<FeaturizeTextOption>();
+
+var pipeline = context.Transforms.Categorical.OneHotEncoding(new[] { new InputOutputColumnPair(@"Sex", @"Sex"), new InputOutputColumnPair(@"Embarked", @"Embarked") })
+                .Append(context.Transforms.Concatenate(@"TextFeature", @"Name", "Ticket", "Cabin"))
+                .Append(context.Sweepable().CreateSweepableEstimator(

```

Comment:
Is there a reason the `Sweepable()` extension uses the parenthesis when none of the other extensions on the MLContext use them?

Reply:
`Sweepable()` is not the only extension that uses parenthesis. AFAIK `Auto()` and `Recommendation` extension also have parenthesis.

Reply:
Ah yeah, I see. Its based on if its an extension method or not.